File: Flutter/App/test/repository/product_repository_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/annotations.dart';
import 'package:mockito/mockito.dart';
import 'package:products_crud/domain/product/model/product.dart';
import 'package:products_crud/domain/product/model/product_details.dart';
import 'package:products_crud/domain/product/repository/product_repository.dart';

import 'product_repository_test.mocks.dart';

@GenerateMocks([ProductRepository])
void main() {
  late MockProductRepository mockProductRepository;

  const mockProductDetails = ProductDetails(
    id: 1,
    title: 'Product 1',
    description: 'Detailed description of Product 1',
    price: 100,
    category: 'Category 1',
    images: ['image1.png'],
  );

  final mockProductList = [
    const Product(
      id: 1,
      title: 'Product 1',
      description: 'Description 1',
      price: 10,
      category: 'Category 1',
      images: ['image1.png'],
    ),
    const Product(
      id: 2,
      title: 'Product 2',
      description: 'Description 2',
      price: 20,
      category: 'Category 2',
      images: ['image2.png'],
    ),
  ];

  setUp(() {
    mockProductRepository = MockProductRepository();
  });

  group('ProductRepository Tests', () {
    test('should return ProductDetails when getProductDetails is called', () async {
      const productId = 1;
      when(mockProductRepository.getProductDetails(id: productId)).thenAnswer((_) async => mockProductDetails);

      final result = await mockProductRepository.getProductDetails(id: productId);

      expect(result, mockProductDetails);
      verify(mockProductRepository.getProductDetails(id: productId)).called(1);
      verifyNoMoreInteractions(mockProductRepository);
    });

    test('should return a list of Products when getProducts is called', () async {
      when(mockProductRepository.getProducts()).thenAnswer((_) async => mockProductList);

      final result = await mockProductRepository.getProducts();

      expect(result, mockProductList);
      verify(mockProductRepository.getProducts()).called(1);
      verifyNoMoreInteractions(mockProductRepository);
    });

    test('should throw an exception when getProductDetails fails', () async {
      const productId = 1;
      when(mockProductRepository.getProductDetails(id: productId))
          .thenThrow(Exception('Failed to fetch product details'));

      expect(
        () async => mockProductRepository.getProductDetails(id: productId),
        throwsA(isA<Exception>()),
      );
      verify(mockProductRepository.getProductDetails(id: productId)).called(1);
      verifyNoMoreInteractions(mockProductRepository);
    });

    test('should throw an exception when getProducts fails', () async {
      when(mockProductRepository.getProducts()).thenThrow(Exception('Failed to fetch products'));

      expect(
        () async => mockProductRepository.getProducts(),
        throwsA(isA<Exception>()),
      );
      verify(mockProductRepository.getProducts()).called(1);
      verifyNoMoreInteractions(mockProductRepository);
    });
  });
}

File: Android/App/app/src/test/java/com/efalco777/crud_products/repository/ProductRepositoryTest.kt
package com.efalco777.crud_products.repository

import com.efalco777.crud_products.domain.product.model.Product
import com.efalco777.crud_products.domain.product.model.ProductDetails
import com.efalco777.crud_products.domain.product.repository.ProductRepository
import io.mockk.*
import junit.framework.TestCase.assertEquals
import kotlinx.coroutines.test.*
import org.junit.Before
import org.junit.Test

class ProductRepositoryTest {

    private lateinit var productRepository: ProductRepository

    @Before
    fun setup() {
        productRepository = mockk()

        coEvery { productRepository.getProductList() } returns listOf(
            Product.empty().copy(id = 0, title = "Product A"),
            Product.empty().copy(id = 1, title = "Product B"),
        )

        coEvery { productRepository.getProductDetails(0) } returns ProductDetails.empty()
            .copy(id = 1, description = "Description A")
    }

    @Test
    fun `getProductList should return a list of products`() = runTest {
        val productList = productRepository.getProductList()

        assertEquals(2, productList.size)
        assertEquals("Product A", productList[0].title)
        assertEquals("Product B", productList[1].title)
    }

    @Test
    fun `getProductDetails should return product details for a valid ID`() = runTest {
        val productDetails = productRepository.getProductDetails(0)

        assertEquals(1, productDetails.id)
        assertEquals("Description A", productDetails.description)
    }

    @Test
    fun `getProductList should return empty list when no products are available`() = runTest {
        coEvery { productRepository.getProductList() } returns emptyList()

        val productList = productRepository.getProductList()

        assertEquals(0, productList.size)
    }
}
